### PR TITLE
gh-95299: Stop installing setuptools as a part of ensurepip and venv

### PR DESCRIPTION
**Pull request by pradyunsg**
This PR removes the bundled setuptools wheel from ensurepip, and stops installing setuptools in environments created by venv.

---

I based this PR off of my understanding of venv and ensurepip, and using the following search to locate all the relevant bits of code + ensuring that tests are happy locally.

Screenshot 2023-01-14 at 17 20 01

In this PR, I've intentionally not touched the example in https://docs.python.org/3.12/library/venv.html#an-example-of-extending-envbuilder. That example is largely outdated and is likely better served being updated in a dedicated PR for it.

* Issue: gh-95299

### CODE REVIEW

**Comment by pradyunsg** [comment]
Hmm... @vstinner @encukou any opinions on what I should do with `test_cppext` (added in https://github.com/python/cpython/issues/91321) here?

That test has an implicit assumption that `setuptools` will be present in the virtual environment created by `venv` (added in https://github.com/python/cpython/pull/92639, as part of removing reliance on distutils).

Notably:

- Does this test need to use packaging tooling, or would it be OK to hand-craft building the extension and rely on that?
- If it _should_ use setuptools, would it be OK for this test to require reaching out over the network to fetch setuptools via pip?
- Would it be OK to build the extension in pip-installable package, rather than [using the deprecated approach of directly _invoking_ setup.py](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html)?

**Comment by pradyunsg** [comment]
Since it's cheap to revert commits/remove them from a PR, and nice to have a code change to discuss details over, I've gone ahead and pushed a change that assumes that the answer to the questions I raised above is yes, yes and yes. :)

File: .github/workflows/verify-ensurepip-wheels.yml
Comment (hugovk) [inline .github/workflows/verify-ensurepip-wheels.yml]: Now singular 
```suggestion
      - name: Compare checksum of bundled pip to the one published on PyPI
```

**Review by vsajip — APPROVED** [review]
The changes seem reasonable to me, though I'm only the code owner for the venv and test_venv bits.

**Comment by encukou** [comment]
> I guess we can hand-craft building the extension and rely on that

IMO, that would be best. Relatedly, I've been thinking how to handle `distutils` removal in [the extension tutorial](https://docs.python.org/3.12/extending/newtypes_tutorial.html) -- it should definitely mention setuptools, but a simple but working example command line, because setuptools is not always available.
Having that in the tests -- even in a way that only satisfies our CI on select platforms -- would help :)
CPython tests depending on setuptools isn't great in general, and neither is downloading code as part of the test run.

Anyway, if that sounds like more than you signed up for, go ahead with Setuptools.
I don't have time for a full PR review now, sorry! Whoever gets around to it, please `test-with-buildbots` before merging.

**Comment by pradyunsg** [comment]
/cc @dstufft @ncoghlan as other experts on ensurepip.

I'm _technically_ listed as an expert on ensurepip already and I'm ~sure that the changes I've made to it are correct; but it certainly won't hurt if y'all can review this. :)

**Review by pfmoore — APPROVED** [review]

File: Lib/test/test_cppext.py
Comment (pfmoore) [inline Lib/test/test_cppext.py]: This is the unadorned directory name, which is treated below by pip as a *package* name, not a directory.

Add an `os.path.abspath()` round this to disambiguate?

**Comment by pfmoore** [comment]
I added a review comment that should fix the buildbot errors.

File: Lib/ensurepip/__init__.py
Comment (merwok) [inline Lib/ensurepip/__init__.py:157]: more mentions of easy_install on lines 161 and 164

File: Lib/test/test_cppext.py
Comment (merwok) [inline Lib/test/test_cppext.py]: I would’ve put `env` right after `cmd`, as they are related to the process model, whereas `text` is python processing.
Very minor of course.

**Comment by hugovk** [comment]
Should this also go in What's New, under [Removed](https://docs.python.org/3.12/whatsnew/3.12.html#removed)?

**Comment by pfmoore** [comment]
> Should this also go in What's New, under [Removed](https://docs.python.org/3.12/whatsnew/3.12.html#removed)?

The inclusion of setuptools was always explicitly documented as an internal implementation detail, and as such, I don't think we have any obligation to report its removal. It could be mentioned in the same way that we might document the removal of an internal API - I did a quick check and the 3.11 What's New included "Removed the undocumented private `float.__set_format__()` method". So I guess there's a precedent.

**Comment by encukou** [comment]
Please add it as a convenience to users, even if there's technically no obligation. It's been *very* easy to end up depending on setuptools accidentally.

IMO, for removals/breaks, it's best to write What's New entries for the target audience of a poor soul who isn't too well-versed in Python but inherited some complex system with a “temporary” hack. Point them to a solution, don't apologize, but don't berate them either.
With that in mind, please mention `pkg_resources`, as that shows up in ImportErrors but the connection to setuptools may not be clear.

(cc @python/proofreaders -- hope the above is good general advice)

**Comment by AlexWaygood** [comment]
As a (somewhat specific) example for how this change can affect users: the bundling of setuptools by default has been playing havoc with typeshed's tests for our stubs. We have a test that dynamically imports every stdlib module at runtime and does various consistency checks against our stubs (checks that parameter names are the same, checks that there are the same number of parameters, etc etc).

The bundling of setuptools by default means that you now get the setuptools version of `distutils` if you do `import distutils` at runtime on Python 3.10+. This means typeshed gets a host of spurious new errors in our test suite every time the bundled version of setuptools is updated in a Python minor version, since a new release of setuptools generally comes with many changes to the setuptools version of `distutils`. Our test can't detect that the changes have been made to the setuptools version of `distutils` rather than the stdlib version of distutils, so reports a host of "consistency errors" between our stubs for the stdlib `distutils` and the setuptools `distutils` at runtime.

See https://github.com/python/typeshed/issues/9520 for more details

**Comment by AlexWaygood** [comment]
> Please don't merge this until those fixups are squashed

All PRs are squash-merged into a single commit in the CPython repo, so no need to worry about that :)

**Comment by encukou** [comment]
In CPython, PRs are always squashed when merging, so squashing fixups isn't necessary. In fact, it's best to avoid force-pushes, since it doesn't play well with some people's workflows. See the [devguide](https://devguide.python.org/getting-started/pull-request-lifecycle/#quick-guide).

Anyway, welcome to the team, glad to have you here!

**Comment by pradyunsg** [comment]
Ah, indeed! In that case, less work for future me! :)

**Comment by pfmoore** [comment]
LGTM (in particular the "What's New" entry). I've already approved so my approval stands 🙂

**Comment by CAM-Gerlach** [comment]
> IMO, for removals/breaks, it's best to write What's New entries for the target audience of a poor soul who isn't too well-versed in Python but inherited some complex system with a “temporary” hack. Point them to a solution, don't apologize, but don't berate them either.
>
> (cc @python/proofreaders -- hope the above is good general advice)

Yup, FWIW I would definitely recommend erring on the side of caution and going with the above, particularly in this case. It's particularly useful as a place to point people to, even if they don't always find it on their own.

File: Doc/whatsnew/3.12.rst
Comment (CAM-Gerlach) [inline Doc/whatsnew/3.12.rst:554]:
```suggestion
  :mod:`venv` or bootstrapped with :mod:`ensurepip`, since they are part of
```

May as well cross-ref these, no?

File: Doc/whatsnew/3.12.rst
Comment (CAM-Gerlach) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* Remove the bundled setuptools wheel from :mod:`ensurepip`,
  and stop installing setuptools in environments created by :mod:`venv`.
```

* Cross-reference the respective modules
* Line wrap

**Comment by CAM-Gerlach** [comment]
There's still a number of references to `setuptools` in the `venv` docs, in the [Creating virtual environments](https://deploy-preview-101039--python-cpython-preview.netlify.app/library/venv.html#creating-virtual-environments) and [An example to extending EnvBuilder](https://deploy-preview-101039--python-cpython-preview.netlify.app/library/venv.html#an-example-of-extending-envbuilder) sections. Should any of those be removed? Seems like at least the former should, as `--upgrade-deps` no longer mentioned `Setuptools` (since it uses `CORE_VENV_DEPS` that you updated).

**Comment by pradyunsg** [comment]
> Should any of those be removed?

Yes. I'll remove the reference from the section about creating venvs. For the example, as I said in the PR description, I've intentionally not touched it.

**Comment by CAM-Gerlach** [comment]
> Yes. I'll remove the reference from the section about creating venvs.

(Don't forget this :) )

**Review by dstufft — APPROVED** [review]

File: Doc/using/venv-create.inc
Comment (merwok) [inline Doc/using/venv-create.inc]:
```suggestion
      --upgrade-deps        Upgrade core dependencies (pip) to the
```

Comment (pradyunsg) [reply Doc/using/venv-create.inc]: Done, and updated the code that this is supposed to be equivalent to as well.

File: .github/workflows/verify-ensurepip-wheels.yml
Comment (merwok) [inline .github/workflows/verify-ensurepip-wheels.yml]: For this file I think it would be fine to not be specific in the names, like:
Verify bundled wheels
Compare checksums of bundled wheels to the ones on PyPI

File: Lib/test/test_cppext.py
Comment (CAM-Gerlach) [inline Lib/test/test_cppext.py]: LGTM on the docs now, thanks.

You you're probably already on it, but seems like this is failing on [some of the buildbots](https://buildbot.python.org/all/#/builders/465/builds/1049) with

```
subprocess.CalledProcessError: Command '['env/bin/python', '-X', 'dev', '-m', 'pip', 'install', '/buildbot/buildarea/cpython/pull_request.ware-gentoo-x86.installed/build/target/lib/python3.12/test/cppextdata']' returned non-zero exit status 1.
```

And reporting the following error:

```
ERROR: Invalid requirement: '/buildbot/buildarea/cpython/pull_request.ware-gentoo-x86.installed/build/target/lib/python3.12/test/cppextdata'
```

Comment (pradyunsg) [reply Lib/test/test_cppext.py]: I'm aware -- these specific buildbots have been consistently failing since the start, hence me triggering runs on them with smaller changes to the test setup. :)

I've not been able to reproduce the error so far outside of those environments, and not had the time/knowledge to investigate what's going on with these buildbots.

Comment (CAM-Gerlach) [reply Lib/test/test_cppext.py]: Okay, thanks. I've dug in a bit and found the following (which you probably already did, but just documenting this for completeness for any others reading this):

* s390x RHEL7 & 8 refleaks failures with `test_asyncio` appears to be pre-existing pre-existing and error identically on other older and newer PRs
* s390x Fedora LTO + PGO failure with missing `make` target likewise fails the same way with older and newer PRs, so evidently also unrelated
* AMD64 Windows11 Bigmem test error with `test_hashlib` does not appear to be exactly consistent with others, some of which passed and some failed with different errors, but it has very little build history to go on, it appears rather flaky (it is bigmem after all) and the test failure here is only on a single run and doesn't seem obviously related, so maybe we can defer it for now.
* The remaining failures all occur if and only if the buildbot is an `Installed` buildbot (all Linux, various distros and platforms), where `pip` apparently cannot find the `cppextdata` directory with the C extension to build at the installed path. Given the search path, `__file__ / "cppextdata"` is different than for the non-installed versions (under `build/target/lib/python3.12/test/cppextdata` rather than `build/Lib/test/cppextdata`, I'm assuming the `cppextdata` is not present in that specific directory when installed, and is instead somewhere else or not installed/copied.

As to where specifically it is installed/copied I'm not sure and I'm not expert on FHS/Linux filesystem layouts, but I would think that's something that can be reproduced after installing in on a Linux box, VM a similar manner (or maybe just asking Filipe who is the expert at Linux install schemes, or perhaps its something that one of the other Linux gurus would know)?

File: Lib/test/test_cppext.py
Comment (vstinner) [inline Lib/test/test_cppext.py]: It would be better to have a decorator like ``@test.support.requires_venv_with_pip``. requires_venv_with_pip() checks also for zlib and ctypes, so it's more complicated than just ssl in practice.

Comment (pradyunsg) [reply Lib/test/test_cppext.py]: We already have that decorator on this test.

**Comment by pradyunsg** [comment]
I've filed #103316, pulling out the test_cppext changes into that PR.

**Comment by pradyunsg** [comment]
OK, and now I've pulled in changes in the changes from that PR and addressed all outstanding concerns. Assuming the CI and buildbots are happy, this should be good to go.

**Comment by pradyunsg** [comment]
All the buildbot failures seem like false negatives, with test_asyncio, test_eintr, test_toolsaand test_typing failing for unrelated reasons.

**Comment by pradyunsg** [comment]
Given that this has multiple reviews, that the CI and buildbots are happy (all the non-flaky ones), that all outstanding comments are addressed and the multiple approvals at various points, I'm going to go ahead and merge this. If there's more needed here, we can handle that in a follow up!

**Comment by pradyunsg** [comment]
Thanks all! Holler on the issue, if there's any concerns! :)

**Pull request merged**
